flink-table/flink-table-planner/src/main/java/org/apache/calcite/rel/rules/MultiJoin.java

```diff
@@ -0,0 +1,306 @@
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import org.apache.calcite.linq4j.Ord;
+import org.apache.calcite.plan.Convention;
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelTraitSet;
+import org.apache.calcite.rel.AbstractRelNode;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.RelWriter;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rel.hint.RelHint;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexShuttle;
+import org.apache.calcite.util.ImmutableBitSet;
+import org.apache.calcite.util.ImmutableIntList;
+import org.apache.calcite.util.ImmutableNullableList;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A MultiJoin represents a join of N inputs, whereas regular Joins represent strictly binary joins.
+ */
+public final class MultiJoin extends AbstractRelNode {
+    // ~ Instance fields --------------------------------------------------------
+
+    private final List<RelNode> inputs;
+    private final RexNode joinFilter;
+
+    @SuppressWarnings("HidingField")
+    private final RelDataType rowType;
+
+    private final boolean isFullOuterJoin;
+    private final List<@Nullable RexNode> outerJoinConditions;
+    private final ImmutableList<JoinRelType> joinTypes;
+    private final List<@Nullable ImmutableBitSet> projFields;
+    public final ImmutableMap<Integer, ImmutableIntList> joinFieldRefCountsMap;
+    private final @Nullable RexNode postJoinFilter;
+    private final ImmutableList<RelHint> hints;
+
+    // ~ Constructors -----------------------------------------------------------
+
+    /**
+     * Constructs a MultiJoin.
+     *
+     * @param cluster cluster that join belongs to
+     * @param hints hints for this multi-join
+     * @param inputs inputs into this multi-join
+     * @param joinFilter join filter applicable to this join node
+     * @param rowType row type of the join result of this node
+     * @param isFullOuterJoin true if the join is a full outer join
```

Comment:
it looks like the tests only cover this way inner joins. I think we should add tests to drive, isFullOuterJoin true and false, with and without outerJoinConditions and with an without projections. With and without TTLs in each case.